# 1. Context

> **TODO — slot not started yet.**
>
> Add here: project objective, business question(s) this EDA supports, and what decisions the pipeline design will depend on.

# 2. Data Sources

In [1]:
import pandas as pd
import plotly.express as px
import plotly.io as pio
pio.renderers.default = "vscode"
import os

raw_path = "../data/raw"

os.listdir(raw_path)

['taxi_trip_data.csv',
 '.gitkeep',
 'taxi_zone_geo.csv',
 '.complete',
 'original_cleaned_nyc_taxi_data_2018.csv']

In [2]:
#Loading the taxi trips .csv
trips = pd.read_csv("../data/raw/taxi_trip_data.csv", nrows=100_000)

#Loading the geographics zones .csv
zones = pd.read_csv("../data/raw/taxi_zone_geo.csv")

#Loading "clean"  data
cleaned = pd.read_csv("../data/raw/original_cleaned_nyc_taxi_data_2018.csv", nrows=100_000)

# 3. Overview of the 3 datasets

## Quick Inspection

In [3]:
def quick(df):
    print("shape:", df.shape)
    print("\ncolumns:")
    print(df.columns)
    print("\ndtypes:")
    print(df.dtypes)
    print("\nisnull:")
    print(df.isnull().sum())
    display(df.head())

## Trips

In [4]:
quick(trips)

shape: (100000, 17)

columns:
Index(['vendor_id', 'pickup_datetime', 'dropoff_datetime', 'passenger_count',
       'trip_distance', 'rate_code', 'store_and_fwd_flag', 'payment_type',
       'fare_amount', 'extra', 'mta_tax', 'tip_amount', 'tolls_amount',
       'imp_surcharge', 'total_amount', 'pickup_location_id',
       'dropoff_location_id'],
      dtype='object')

dtypes:
vendor_id                int64
pickup_datetime         object
dropoff_datetime        object
passenger_count          int64
trip_distance          float64
rate_code                int64
store_and_fwd_flag      object
payment_type             int64
fare_amount            float64
extra                  float64
mta_tax                float64
tip_amount             float64
tolls_amount           float64
imp_surcharge          float64
total_amount           float64
pickup_location_id       int64
dropoff_location_id      int64
dtype: object

isnull:
vendor_id              0
pickup_datetime        0
dropoff_datetime     

,vendor_id,pickup_datetime,dropoff_datetime,passenger_count,trip_distance,rate_code,store_and_fwd_flag,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,imp_surcharge,total_amount,pickup_location_id,dropoff_location_id
0,2,2018-03-29 13:37:13,2018-03-29 14:17:01,1,18.15,3,N,1,70.0,0.0,0.0,16.16,10.50,0.3,96.96,161,1
1,2,2018-03-29 13:37:18,2018-03-29 14:15:33,1,4.59,1,N,1,25.0,0.0,0.5,5.16,0.00,0.3,30.96,13,230
2,2,2018-03-29 13:26:57,2018-03-29 13:28:03,1,0.30,1,N,1,3.0,0.0,0.5,0.76,0.00,0.3,4.56,231,231
3,2,2018-03-29 13:07:48,2018-03-29 14:03:05,2,16.97,1,N,1,49.5,0.0,0.5,5.61,5.76,0.3,61.67,231,138
4,2,2018-03-29 14:19:11,2018-03-29 15:19:59,5,14.45,1,N,1,45.5,0.0,0.5,10.41,5.76,0.3,62.47,87,138


## Zones

In [5]:
quick(zones)

shape: (263, 4)

columns:
Index(['zone_id', 'zone_name', 'borough', 'zone_geom'], dtype='object')

dtypes:
zone_id       int64
zone_name    object
borough      object
zone_geom    object
dtype: object

isnull:
zone_id      0
zone_name    0
borough      0
zone_geom    0
dtype: int64


,zone_id,zone_name,borough,zone_geom
0,1,Newark Airport,EWR,"POLYGON((-74.1856319999999 40.6916479999999, -..."
1,3,Allerton/Pelham Gardens,Bronx,"POLYGON((-73.848596761 40.8716707849999, -73.8..."
2,18,Bedford Park,Bronx,"POLYGON((-73.8844286139999 40.8668003789999, -..."
3,20,Belmont,Bronx,"POLYGON((-73.8839239579998 40.8644177609999, -..."
4,31,Bronx Park,Bronx,"POLYGON((-73.8710017319999 40.8572767429999, -..."


## Cleaned

In [6]:
quick(cleaned)

shape: (100000, 21)

columns:
Index(['Unnamed: 0', 'trip_distance', 'rate_code', 'store_and_fwd_flag',
       'payment_type', 'fare_amount', 'extra', 'mta_tax', 'tip_amount',
       'tolls_amount', 'imp_surcharge', 'total_amount', 'pickup_location_id',
       'dropoff_location_id', 'year', 'month', 'day', 'day_of_week',
       'hour_of_day', 'trip_duration', 'calculated_total_amount'],
      dtype='object')

dtypes:
Unnamed: 0                   int64
trip_distance              float64
rate_code                    int64
store_and_fwd_flag          object
payment_type                 int64
fare_amount                float64
extra                      float64
mta_tax                    float64
tip_amount                 float64
tolls_amount               float64
imp_surcharge              float64
total_amount               float64
pickup_location_id           int64
dropoff_location_id          int64
year                         int64
month                        int64
day                 

,Unnamed: 0,trip_distance,rate_code,store_and_fwd_flag,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,...,total_amount,pickup_location_id,dropoff_location_id,year,month,day,day_of_week,hour_of_day,trip_duration,calculated_total_amount
0,3,16.97,1,N,1,49.5,0.0,0.5,5.61,5.76,...,61.67,231,138,2018,3,29,3,13,3317.0,61.67
1,4,14.45,1,N,1,45.5,0.0,0.5,10.41,5.76,...,62.47,87,138,2018,3,29,3,14,3648.0,62.47
2,5,11.60,1,N,1,42.0,0.0,0.5,14.57,5.76,...,63.13,68,138,2018,3,29,3,14,3540.0,63.13
3,10,5.10,1,N,1,26.5,1.0,0.5,5.65,0.00,...,33.95,186,33,2018,3,29,3,16,2585.0,33.95
4,12,11.11,1,N,1,45.5,1.0,0.5,10.61,5.76,...,63.67,163,138,2018,3,29,3,16,4521.0,63.67


## General view

In [7]:
datasets = {
    "trips": trips,
    "zones": zones,
    "cleaned": cleaned
}

for name, df in datasets.items():
    print(f"\n===== {name} =====")
    print("Shape:", df.shape)
    print("Columns:", df.columns.tolist())
    print("\nTypes:")
    print(df.dtypes)


===== trips =====
Shape: (100000, 17)
Columns: ['vendor_id', 'pickup_datetime', 'dropoff_datetime', 'passenger_count', 'trip_distance', 'rate_code', 'store_and_fwd_flag', 'payment_type', 'fare_amount', 'extra', 'mta_tax', 'tip_amount', 'tolls_amount', 'imp_surcharge', 'total_amount', 'pickup_location_id', 'dropoff_location_id']

Types:
vendor_id                int64
pickup_datetime         object
dropoff_datetime        object
passenger_count          int64
trip_distance          float64
rate_code                int64
store_and_fwd_flag      object
payment_type             int64
fare_amount            float64
extra                  float64
mta_tax                float64
tip_amount             float64
tolls_amount           float64
imp_surcharge          float64
total_amount           float64
pickup_location_id       int64
dropoff_location_id      int64
dtype: object

===== zones =====
Shape: (263, 4)
Columns: ['zone_id', 'zone_name', 'borough', 'zone_geom']

Types:
zone_id       int64

We are choosing the "taxi_trip_data" as our main source of information because is the .csv that only has information about the taxi trips, and that is what we are going to analyse in this project

# 4. Relationship between datasets

> **TODO — slot not started yet.**
>
> Explore how `taxi_trip_data` relates to `taxi_zone_geo` (via `pickup_location_id` / `dropoff_location_id`) and how `cleaned` compares to the raw `taxi_trip_data` (same source, pre-cleaned reference?).

# 5. EDA - taxi_trip_data.csv

## 5.1 Structure and types

In [8]:
import pandas as pd

#Loading the taxi trips .csv
trips = pd.read_csv("../data/raw/taxi_trip_data.csv", nrows=100_000)

In [9]:
trips.shape

(100000, 17)

In [10]:
trips.head()

,vendor_id,pickup_datetime,dropoff_datetime,passenger_count,trip_distance,rate_code,store_and_fwd_flag,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,imp_surcharge,total_amount,pickup_location_id,dropoff_location_id
0,2,2018-03-29 13:37:13,2018-03-29 14:17:01,1,18.15,3,N,1,70.0,0.0,0.0,16.16,10.50,0.3,96.96,161,1
1,2,2018-03-29 13:37:18,2018-03-29 14:15:33,1,4.59,1,N,1,25.0,0.0,0.5,5.16,0.00,0.3,30.96,13,230
2,2,2018-03-29 13:26:57,2018-03-29 13:28:03,1,0.30,1,N,1,3.0,0.0,0.5,0.76,0.00,0.3,4.56,231,231
3,2,2018-03-29 13:07:48,2018-03-29 14:03:05,2,16.97,1,N,1,49.5,0.0,0.5,5.61,5.76,0.3,61.67,231,138
4,2,2018-03-29 14:19:11,2018-03-29 15:19:59,5,14.45,1,N,1,45.5,0.0,0.5,10.41,5.76,0.3,62.47,87,138


In [11]:
trips.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 17 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   vendor_id            100000 non-null  int64  
 1   pickup_datetime      100000 non-null  object 
 2   dropoff_datetime     100000 non-null  object 
 3   passenger_count      100000 non-null  int64  
 4   trip_distance        100000 non-null  float64
 5   rate_code            100000 non-null  int64  
 6   store_and_fwd_flag   100000 non-null  object 
 7   payment_type         100000 non-null  int64  
 8   fare_amount          100000 non-null  float64
 9   extra                100000 non-null  float64
 10  mta_tax              100000 non-null  float64
 11  tip_amount           100000 non-null  float64
 12  tolls_amount         100000 non-null  float64
 13  imp_surcharge        100000 non-null  float64
 14  total_amount         100000 non-null  float64
 15  pickup_location_id

"pickup_datetime" and "dropoff_datetime" as object is "wrong" i must change the type for 'date' in the pipeline

payment_type as a number, i found in the description of the dataset the equivalent numbers, so in the future i will need to create a auxiliary table

rate_code is categoric besides beeing int

store_and_fwd_flg is a object but is categoric

In [12]:
schema = pd.DataFrame({
    "column": trips.columns,
    "dtype": trips.dtypes.values,
    "null_count": trips.isnull().sum().values,
    "unique_count": trips.nunique().values
})

schema

,column,dtype,null_count,unique_count
0,vendor_id,int64,0,3
1,pickup_datetime,object,0,99618
2,dropoff_datetime,object,0,99572
3,passenger_count,int64,0,10
4,trip_distance,float64,0,2836
5,rate_code,int64,0,5
6,store_and_fwd_flag,object,0,2
7,payment_type,int64,0,4
8,fare_amount,float64,0,625
9,extra,float64,0,13


17 columns
100,000 records in the sample
0 nulls across all columns in the sample
8 float64
6 int64
3 object

## 5.2 Data quality

In [13]:
trips.isnull().sum()

vendor_id              0
pickup_datetime        0
dropoff_datetime       0
passenger_count        0
trip_distance          0
rate_code              0
store_and_fwd_flag     0
payment_type           0
fare_amount            0
extra                  0
mta_tax                0
tip_amount             0
tolls_amount           0
imp_surcharge          0
total_amount           0
pickup_location_id     0
dropoff_location_id    0
dtype: int64

In [14]:
trips.duplicated().sum()

np.int64(134)

In [15]:
trips['payment_type'].value_counts()

payment_type
1    95188
2     4434
3      291
4       87
Name: count, dtype: int64

In [16]:
trips["vendor_id"].value_counts()

vendor_id
2    60283
1    39250
4      467
Name: count, dtype: int64

In [17]:
trips["rate_code"].value_counts()

rate_code
1    95272
5     2358
3     1889
4      473
2        8
Name: count, dtype: int64

In [18]:
trips[trips["trip_distance"] <= 0].count()

vendor_id              1899
pickup_datetime        1899
dropoff_datetime       1899
passenger_count        1899
trip_distance          1899
rate_code              1899
store_and_fwd_flag     1899
payment_type           1899
fare_amount            1899
extra                  1899
mta_tax                1899
tip_amount             1899
tolls_amount           1899
imp_surcharge          1899
total_amount           1899
pickup_location_id     1899
dropoff_location_id    1899
dtype: int64

In [19]:
trips[trips ['pickup_location_id'] == trips['dropoff_location_id']].count()

vendor_id              4647
pickup_datetime        4647
dropoff_datetime       4647
passenger_count        4647
trip_distance          4647
rate_code              4647
store_and_fwd_flag     4647
payment_type           4647
fare_amount            4647
extra                  4647
mta_tax                4647
tip_amount             4647
tolls_amount           4647
imp_surcharge          4647
total_amount           4647
pickup_location_id     4647
dropoff_location_id    4647
dtype: int64

Doing this data quality analyse we can see some errors:
1) Found 134 lines duplicates
2) Found 1899 lines with trip_distance <= 0
3) Found 4647 lines with pickup_location_id == dropoff_location_id

Now we will analyse this lines to understand whats happening

In [20]:
trips_wrongs = trips[trips["trip_distance"] <= 0]
trips_wrongs.head()

,vendor_id,pickup_datetime,dropoff_datetime,passenger_count,trip_distance,rate_code,store_and_fwd_flag,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,imp_surcharge,total_amount,pickup_location_id,dropoff_location_id
87,1,2018-05-09 08:37:22,2018-05-09 08:37:43,2,0.0,5,N,1,60.0,0.0,0.0,12.05,0.0,0.3,72.35,132,132
168,2,2018-02-01 19:45:04,2018-02-01 19:45:06,2,0.0,5,N,1,186.5,0.0,0.5,37.46,0.0,0.3,224.76,264,265
178,2,2018-02-01 23:46:19,2018-02-01 23:46:25,1,0.0,5,N,1,156.0,0.0,0.0,31.26,0.0,0.3,187.56,264,265
231,2,2018-07-12 12:48:21,2018-07-12 12:53:30,1,0.0,1,N,1,2.5,0.0,0.5,16.00,0.0,0.3,19.30,68,68
297,2,2018-12-15 19:55:01,2018-12-15 19:55:06,1,0.0,1,N,1,2.5,0.0,0.5,0.66,0.0,0.3,3.96,264,193


In [21]:
trips_wrongs.shape

(1899, 17)

In [22]:
duplicates = trips[trips.duplicated(keep=False)]
#keep = false, show from the first apparence to the final, if we keep=True, will only show from the second ocorrence

In [23]:
duplicates.head()

,vendor_id,pickup_datetime,dropoff_datetime,passenger_count,trip_distance,rate_code,store_and_fwd_flag,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,imp_surcharge,total_amount,pickup_location_id,dropoff_location_id
1314,1,2018-03-13 16:52:05,2018-03-13 17:32:14,1,8.20,1,N,1,30.0,1.0,0.5,6.35,0.00,0.3,38.15,50,17
2724,1,2018-03-31 20:40:55,2018-03-31 21:02:33,1,6.70,1,N,1,22.0,0.5,0.5,4.65,0.00,0.3,27.95,264,82
3118,2,2018-03-13 23:17:34,2018-03-13 23:34:06,1,9.38,1,N,1,27.0,0.5,0.5,4.00,0.00,0.3,32.30,138,144
3216,1,2018-03-01 17:53:05,2018-03-01 19:15:23,2,15.30,1,N,1,56.5,1.0,0.5,10.00,5.76,0.3,74.06,264,264
3249,2,2018-03-22 23:04:46,2018-03-22 23:35:57,5,16.86,1,N,1,47.5,0.5,0.5,7.32,0.00,0.3,56.12,132,255


In [24]:
duplicates.shape

(268, 17)

In [25]:
duplicates.sort_values(
    by=["pickup_datetime", "dropoff_datetime"]
).head(10)

,vendor_id,pickup_datetime,dropoff_datetime,passenger_count,trip_distance,rate_code,store_and_fwd_flag,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,imp_surcharge,total_amount,pickup_location_id,dropoff_location_id
6777,2,2018-03-01 00:53:19,2018-03-01 01:09:54,2,7.33,1,N,1,22.0,0.5,0.5,4.66,0.00,0.3,27.96,209,263
28936,2,2018-03-01 00:53:19,2018-03-01 01:09:54,2,7.33,1,N,1,22.0,0.5,0.5,4.66,0.00,0.3,27.96,209,263
4754,2,2018-03-01 00:58:35,2018-03-01 01:25:34,1,19.83,1,N,1,52.5,0.5,0.5,9.00,0.00,0.3,62.80,132,49
27911,2,2018-03-01 00:58:35,2018-03-01 01:25:34,1,19.83,1,N,1,52.5,0.5,0.5,9.00,0.00,0.3,62.80,132,49
8858,2,2018-03-01 03:47:33,2018-03-01 03:48:28,2,0.31,1,N,1,3.0,0.5,0.5,1.08,0.00,0.3,5.38,137,137
33095,2,2018-03-01 03:47:33,2018-03-01 03:48:28,2,0.31,1,N,1,3.0,0.5,0.5,1.08,0.00,0.3,5.38,137,137
25690,2,2018-03-01 16:24:07,2018-03-01 17:19:05,1,16.57,1,N,1,50.5,1.0,0.5,11.61,5.76,0.3,69.67,138,13
49291,2,2018-03-01 16:24:07,2018-03-01 17:19:05,1,16.57,1,N,1,50.5,1.0,0.5,11.61,5.76,0.3,69.67,138,13
3216,1,2018-03-01 17:53:05,2018-03-01 19:15:23,2,15.30,1,N,1,56.5,1.0,0.5,10.00,5.76,0.3,74.06,264,264
41226,1,2018-03-01 17:53:05,2018-03-01 19:15:23,2,15.30,1,N,1,56.5,1.0,0.5,10.00,5.76,0.3,74.06,264,264


I identified 134 duplicated lines, we can see that all columns are duplicates so probably are really duplicates but to have more confidence i will run another lines to analyse especifically other columns

Running another checkup i can confirm that is really duplicates, because we can see that ALL columns are duplicates too

In the 100.000 we found 134 duplicates that translate to 268 lines

So in the pipeline this duplicates will be removed

In [26]:
duplicate_rows = len(duplicates)
percentage = duplicate_rows / len(trips) * 100

print(f"Rows involved in duplication: {duplicate_rows}")
print(f"Percentage: {percentage:.3f}%")

Rows involved in duplication: 268
Percentage: 0.268%


In [27]:
duplicate_counts = trips.value_counts()

duplicate_counts[duplicate_counts > 1].head(20)

vendor_id  pickup_datetime      dropoff_datetime     passenger_count  trip_distance  rate_code  store_and_fwd_flag  payment_type  fare_amount  extra  mta_tax  tip_amount  tolls_amount  imp_surcharge  total_amount  pickup_location_id  dropoff_location_id
2          2018-03-07 07:15:15  2018-03-07 07:39:43  2                8.56           1          N                   1             27.5         0.0    0.5      6.81        5.76          0.3            40.87         138                 236                    2
           2018-03-02 20:56:02  2018-03-02 21:47:55  1                25.40          1          N                   1             69.5         0.5    0.5      18.24       2.16          0.3            91.20         90                  201                    2
1          2018-03-17 15:20:02  2018-03-17 16:06:17  2                5.30           1          N                   1             29.5         0.0    0.5      9.05        0.00          0.3            39.35         79            

In [28]:
trips_wrongs = trips[trips["trip_distance"] <= 0]
trips_wrongs.head()

,vendor_id,pickup_datetime,dropoff_datetime,passenger_count,trip_distance,rate_code,store_and_fwd_flag,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,imp_surcharge,total_amount,pickup_location_id,dropoff_location_id
87,1,2018-05-09 08:37:22,2018-05-09 08:37:43,2,0.0,5,N,1,60.0,0.0,0.0,12.05,0.0,0.3,72.35,132,132
168,2,2018-02-01 19:45:04,2018-02-01 19:45:06,2,0.0,5,N,1,186.5,0.0,0.5,37.46,0.0,0.3,224.76,264,265
178,2,2018-02-01 23:46:19,2018-02-01 23:46:25,1,0.0,5,N,1,156.0,0.0,0.0,31.26,0.0,0.3,187.56,264,265
231,2,2018-07-12 12:48:21,2018-07-12 12:53:30,1,0.0,1,N,1,2.5,0.0,0.5,16.00,0.0,0.3,19.30,68,68
297,2,2018-12-15 19:55:01,2018-12-15 19:55:06,1,0.0,1,N,1,2.5,0.0,0.5,0.66,0.0,0.3,3.96,264,193


In [29]:
trips_wrongs.shape

(1899, 17)

In [30]:
trips_wrongs_rows = len(trips_wrongs)
percentage = trips_wrongs_rows / len(trips) * 100

print(f"Rows involved in trip_distance <= 0: {trips_wrongs_rows}")
print(f"Percentage: {percentage:.3f}%")

Rows involved in trip_distance <= 0: 1899
Percentage: 1.899%


1899 in 100.000, we talking about less than 2% so we can delete this amount with no deal

In [31]:
trips_location_wrong = trips[trips['pickup_location_id'] == trips["dropoff_location_id"]]

trips_location_wrong.shape

(4647, 17)

In [32]:
trips_location_wrong[trips_location_wrong['trip_distance'] <= 0].shape

(1357, 17)

In [33]:
trips_location_wrong[trips_location_wrong['trip_distance'] < 0].shape

(0, 17)

In [34]:
trips_location_wrong.head()

,vendor_id,pickup_datetime,dropoff_datetime,passenger_count,trip_distance,rate_code,store_and_fwd_flag,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,imp_surcharge,total_amount,pickup_location_id,dropoff_location_id
2,2,2018-03-29 13:26:57,2018-03-29 13:28:03,1,0.30,1,N,1,3.0,0.0,0.5,0.76,0.00,0.3,4.56,231,231
20,1,2018-03-29 17:07:32,2018-03-29 18:13:31,2,14.10,1,N,1,48.5,1.0,0.5,7.00,5.76,0.3,63.06,264,264
32,2,2018-05-19 11:39:41,2018-05-19 11:40:57,1,0.27,1,N,1,3.0,0.0,0.5,1.14,0.00,0.3,4.94,151,151
60,2,2018-03-12 18:24:53,2018-03-12 18:26:20,1,0.25,1,N,1,3.0,1.0,0.5,0.96,0.00,0.3,5.76,107,107
87,1,2018-05-09 08:37:22,2018-05-09 08:37:43,2,0.00,5,N,1,60.0,0.0,0.0,12.05,0.00,0.3,72.35,132,132


In [35]:
trips_no_distance = trips_location_wrong[trips_location_wrong['trip_distance'] == 0]
trips_no_distance.shape

(1357, 17)

In [36]:
trips_no_distance[trips_no_distance["fare_amount"] == 0].shape

(11, 17)

In [37]:
trips_no_distance[trips_no_distance["fare_amount"] < 0].shape

(20, 17)

In [38]:
fare_invalid = trips_no_distance[
    trips_no_distance["fare_amount"] <= 0
]

fare_invalid[[
    "pickup_datetime",
    "dropoff_datetime",
    "trip_distance",
    "fare_amount",
    "extra",
    "mta_tax",
    "tip_amount",
    "tolls_amount",
    "total_amount",
    "payment_type",
    "rate_code"
]]

,pickup_datetime,dropoff_datetime,trip_distance,fare_amount,extra,mta_tax,tip_amount,tolls_amount,total_amount,payment_type,rate_code
9196,2018-07-23 20:05:49,2018-07-23 20:05:55,0.0,-2.5,-0.5,-0.5,-0.54,0.00,-4.34,3,1
15476,2018-03-01 16:57:10,2018-03-01 16:58:24,0.0,-2.5,-1.0,-0.5,-0.86,0.00,-5.16,3,1
43582,2018-06-22 11:29:33,2018-06-22 11:30:12,0.0,0.0,0.0,0.0,16.88,0.00,17.18,1,5
52754,2018-03-25 01:21:36,2018-03-25 01:21:41,0.0,-2.5,-0.5,-0.5,-1.14,0.00,-4.94,3,1
60168,2018-04-03 11:48:44,2018-04-03 11:49:24,0.0,0.0,0.0,0.0,67.00,0.00,67.30,1,5
74524,2018-06-16 19:37:40,2018-06-16 19:39:46,0.0,0.0,0.0,0.0,277.88,0.00,278.18,1,5
76133,2018-10-18 19:07:55,2018-10-19 17:06:10,0.0,0.0,0.0,0.0,0.00,0.00,0.00,1,1
76240,2018-01-20 18:19:42,2018-01-20 18:19:42,0.0,0.0,0.0,0.0,0.00,0.00,0.00,2,5
76332,2018-08-10 16:55:31,2018-08-10 16:56:06,0.0,-2.5,-1.0,-0.5,0.00,0.00,-4.30,3,1
76367,2018-12-12 01:02:52,2018-12-12 01:08:41,0.0,-2.5,-0.5,-0.5,0.00,0.00,-3.80,3,1


We can see 11 lines where 'fare_amount' = 0 and 20 lines where 'fare_amount' < 0, and doing a more acurated analyse we can see that we have 'total_amount' with negatives and zero values, so lets check this

In [39]:
trips[trips['total_amount'] <= 0].shape

(121, 17)

In [40]:
trips[trips['total_amount'] == 0].shape

(21, 17)

In [41]:
trips[trips['total_amount'] < 0].shape

(100, 17)

In [42]:
total_amount_invalid = trips[
    trips["total_amount"] <= 0
]

total_amount_invalid[[
    "pickup_datetime",
    "dropoff_datetime",
    "trip_distance",
    "fare_amount",
    "extra",
    "mta_tax",
    "tip_amount",
    "tolls_amount",
    "total_amount",
    "payment_type",
    "rate_code"
]]

,pickup_datetime,dropoff_datetime,trip_distance,fare_amount,extra,mta_tax,tip_amount,tolls_amount,total_amount,payment_type,rate_code
2545,2018-06-11 12:50:56,2018-06-11 12:53:35,0.35,-4.0,0.0,-0.5,-0.96,0.0,-5.76,3,1
9196,2018-07-23 20:05:49,2018-07-23 20:05:55,0.00,-2.5,-0.5,-0.5,-0.54,0.0,-4.34,3,1
11306,2018-07-23 19:26:39,2018-07-23 19:35:41,0.89,-7.5,-1.0,-0.5,-80.00,0.0,-89.30,4,1
15476,2018-03-01 16:57:10,2018-03-01 16:58:24,0.00,-2.5,-1.0,-0.5,-0.86,0.0,-5.16,3,1
18966,2018-03-02 18:16:48,2018-03-02 18:16:50,0.00,-2.5,-1.0,-0.5,-0.86,0.0,-5.16,3,1
...,...,...,...,...,...,...,...,...,...,...,...
99728,2018-06-06 20:44:31,2018-06-06 20:45:12,0.00,0.0,0.0,0.0,0.00,0.0,0.00,1,1
99737,2018-02-09 09:16:23,2018-02-09 09:16:33,0.00,-2.5,0.0,-0.5,0.00,0.0,-3.30,4,1
99839,2018-10-18 14:29:37,2018-10-18 14:54:42,3.10,0.0,0.0,0.0,0.00,0.0,0.00,1,1
99866,2018-11-08 18:00:06,2018-11-08 18:04:20,0.06,-4.5,-1.0,-0.5,0.00,0.0,-6.30,4,1


We can see a correlation about total_amount beeing negative with other columns that compose the total_amount formula beeing negative too

## 5.3 Time series analysis

In [43]:
trips_temp = trips.copy()

trips_temp["pickup_datetime"] = pd.to_datetime(
    trips_temp["pickup_datetime"]
)

trips_temp["dropoff_datetime"] = pd.to_datetime(
    trips_temp["dropoff_datetime"]
)

trips_temp[["pickup_datetime", "dropoff_datetime"]].dtypes

pickup_datetime     datetime64[ns]
dropoff_datetime    datetime64[ns]
dtype: object

In [44]:
trips_temp["pickup_datetime"].min()

Timestamp('2009-01-01 00:27:43')

In [45]:
trips_temp["pickup_datetime"].max()

Timestamp('2018-12-31 23:48:14')

In [46]:
trips_temp["dropoff_datetime"].min()

Timestamp('2009-01-01 01:16:52')

In [47]:
trips_temp["dropoff_datetime"].max()

Timestamp('2019-01-01 00:13:35')

The sample covers the period defined by the minimum and maximum values observed in `pickup_datetime` and `dropoff_datetime`.

In [48]:
trips_temp["trip_duration"] = (
    trips_temp["dropoff_datetime"]
    - trips_temp["pickup_datetime"]
)

trips_temp["trip_duration_minutes"] = (
    trips_temp["trip_duration"].dt.total_seconds() / 60
)

trips_temp["trip_duration_minutes"].describe()

count    100000.000000
mean         36.271161
std          78.399632
min         -30.500000
25%          23.900000
50%          30.716667
75%          38.633333
max        1439.366667
Name: trip_duration_minutes, dtype: float64

We can see some errors because how a trip has -30 minutes? So now we will explore more in this

In [49]:
(trips_temp["trip_duration_minutes"] < 0).sum()

np.int64(2)

We have 2 trips that have a duration less than 0 minutes

In [50]:
(trips_temp["trip_duration_minutes"] == 0).sum()

np.int64(92)

We have 92 trips that have a duration equal 0 minutes

First lets see what is in the negative time data lines

In [51]:
negative_duration = trips_temp[trips_temp["trip_duration_minutes"] < 0]

negative_duration[
    [
        "pickup_datetime",
        "dropoff_datetime",
        "trip_duration_minutes",
        "trip_distance",
        "fare_amount",
        "pickup_location_id",
        "dropoff_location_id"
    ]
]

,pickup_datetime,dropoff_datetime,trip_duration_minutes,trip_distance,fare_amount,pickup_location_id,dropoff_location_id
35608,2018-11-04 01:38:13,2018-11-04 01:07:43,-30.50,6.70,26.0,113,37
35610,2018-11-04 01:36:34,2018-11-04 01:08:19,-28.25,9.89,31.5,163,61


### Trip Duration Findings

Trip duration was calculated from the difference between `dropoff_datetime`
and `pickup_datetime`.

The sample contains:
- 2 trips with negative duration;
- 92 trips with zero duration.

Negative durations are considered temporally inconsistent because the
drop-off timestamp occurs before the pick-up timestamp. Since there is not
enough evidence to safely correct these timestamps, these records will be
treated as invalid and removed during the pipeline transformation stage.

Zero-duration trips require contextual analysis because a zero duration
does not necessarily imply that the entire trip record is invalid.

In [52]:
trips_temp["pickup_hour"] = trips_temp["pickup_datetime"].dt.hour

trips_by_hour = (
    trips_temp["pickup_hour"]
    .value_counts()
    .sort_index()
)

trips_by_hour

pickup_hour
0     3600
1     1958
2     1301
3     1003
4     1209
5     1337
6     2022
7     3210
8     4229
9     4581
10    4805
11    4931
12    5207
13    5246
14    5542
15    5589
16    5235
17    5575
18    5558
19    5070
20    5321
21    6194
22    6092
23    5185
Name: count, dtype: int64

In [59]:
fig = px.bar(
    x=trips_by_hour.index,
    y=trips_by_hour.values,
    labels={"x": "Hour of day", "y": "Number of trips"},
    title="Trips by pickup hour",
    text=trips_by_hour.values
)

fig.show()

Pickup volume follows a clear daily cycle. The lowest activity occurs
between 2 AM and 5 AM (roughly 1,000-1,300 trips/hour), consistent with
overnight low demand. Volume ramps up steadily from 6 AM onward, reaching
a stable elevated plateau (~5,200-5,600 trips/hour) from around noon
through the evening. The two highest hours are 9 PM and 10 PM
(6,194 and 6,092 trips respectively), likely reflecting evening
leisure/nightlife demand.

In [63]:
day_order = ["Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday"]

trips_temp["pickup_day_of_week"] = trips_temp["pickup_datetime"].dt.day_name()

trips_by_day = trips_temp["pickup_day_of_week"].value_counts()
trips_by_day = trips_by_day.reindex(day_order)

trips_by_day

pickup_day_of_week
Monday       13401
Tuesday      14939
Wednesday    15662
Thursday     17124
Friday       15269
Saturday     11808
Sunday       11797
Name: count, dtype: int64

In [67]:
fig = px.bar(
    x=trips_by_day.index,
    y=trips_by_day.values,
    labels={"x": "Day of the week", "y": "Number of trips"},
    title="Trips by days of the week",
    text=trips_by_day.values
)

fig.show()

Trip volume is clearly weekday-dominated, peaking on Thursday (17,124 trips)
and staying elevated Tuesday through Friday. Both weekend days are lower
and close to each other (Saturday: 11,808; Sunday: 11,797), roughly 30%
below the Thursday peak. This weekday-skewed pattern — combined with the
evening peak seen in the hourly breakdown (9-10 PM) — suggests the sample
is dominated by commute/after-work trips rather than weekend leisure travel.

## 5.4 Distributions

> **TODO — slot not started yet.**
>
> Histograms / boxplots for `trip_distance`, `fare_amount`, `trip_duration_minutes`, `tip_amount`, `total_amount`. Check skewness and whether a log scale is needed.

## 5.5 Outliers

> **Partially covered above.** Outliers already surfaced during 5.2 (Data quality) and 5.3 (Time series analysis):
>  - `trip_distance <= 0` → 1,899 rows (~1.9%)
>  - `pickup_location_id == dropoff_location_id` → 4,647 rows
>  - `trip_duration_minutes < 0` → 2 rows
>  - `trip_duration_minutes == 0` → 92 rows
>
> **TODO — remaining work for this slot:**
>
> - Consolidate the outlier findings above into one summary table
> - IQR / z-score based outlier detection for `fare_amount`, `trip_distance`, `trip_duration_minutes` (the checks above were rule-based, not statistical)
> - Boxplots for the numeric columns to visualize the outliers

## 5.6 Relationships between variables

> **TODO — slot not started yet.**
>
> Correlation matrix for numeric columns; scatter plots such as `trip_distance` × `fare_amount`, `trip_duration_minutes` × `trip_distance`; check if `payment_type` relates to `tip_amount`.

# 6. EDA - taxi_zone_geo

## 6.1 Structure

> **TODO — slot not started yet.**
>
> `shape`, `head`, `info`, dtypes for `zones` (beyond the initial `quick(zones)` check in section 3).

## 6.2 Quality

> **TODO — slot not started yet.**
>
> Nulls, duplicates, and value ranges for the zone geography fields.

## 6.3 Zone integrity

> **TODO — slot not started yet.**
>
> Check `location_id` is unique / a valid primary key, no orphan or duplicated zone ids.

# 7. Relationship: Trips × Zones

## 7.1 Pickup zones

> **TODO — slot not started yet.**
>
> Distribution of trips by `pickup_location_id`, top/bottom zones by volume.

## 7.2 Dropoff zones

> **TODO — slot not started yet.**
>
> Same analysis as 7.1, but for `dropoff_location_id`.

## 7.3 Unmatched records

> **TODO — slot not started yet.**
>
> Check whether every `pickup_location_id` / `dropoff_location_id` in `trips` exists in `zones` (anti-join); quantify unmatched rows.

# 8. Conclusions

> **TODO — slot not started yet.**
>
> Summarize the key findings from sections 5–7 into a short narrative once they are complete.

# 9. Pipeline requirements

> **Draft — based on findings so far, to be finalized after sections 5.4–8 are complete.**
>
> - Convert `pickup_datetime` and `dropoff_datetime` from object to datetime
> - Create an auxiliary lookup table for `payment_type` codes
> - Treat `rate_code` and `store_and_fwd_flag` as categorical
> - Remove exact duplicate rows (134 duplicates / 268 rows in the 100k sample)
> - Filter out rows with `trip_distance <= 0` (~1.9% of the 100k sample)
> - Remove rows with negative `trip_duration_minutes` (pickup_datetime > dropoff_datetime)
> - Decide how to handle rows with `trip_duration_minutes == 0` (92 rows) — pending investigation
> - Decide how to handle `pickup_location_id == dropoff_location_id` rows (4,647 rows) — pending investigation